### ISSUE TITLE
[Bug]: Fetcher Revalidation Race Condition (Remix)

### ISSUE THREAD

**Issue by brophdawg11** [remix-run/react-router#10473]
### What version of React Router are you using?

6.11.0

### Steps to Reproduce

https://discord.com/channels/@me/1105622443453857842/1105623069365649458

It seems there's a race condition in Remix when the `fetcher.load` kicks off on initial render, and then the subsequent render-driven `navigate` triggers a `shouldRevalidate` call on the fetcher - and if the fetcher has not yet loaded the route module for the Remix route it bombs out.  

The fetcher probably shouldn't attempt to revalidate at all since it's tied to a hardcoded route.  With navigational loaders even on reused routes we still want to call `shouldRevalidate` so folks can opt-in, but with fetchers they can manually re-trigger via `fetcher.load` if necessary.

I do worry this fix might still have the same race condition bug if the render-driven navigation comes from `useSubmit`- since fetchers should attempt to revalidate after actions.  But really GET-driven initial renders shouldn't trigger mutations.  Worth digging a bit deeper into the Remix side of things to see if there's any avenues to look into there.

### Expected Behavior

Fetcher does not try to revalidate and does not call `shouldRevalidate`

### Actual Behavior

Fetcher is trying to revalidate before the route module loads.

**Comment by cmerrick** [comment]
@brophdawg11 I believe we're seeing this in our application, but the discord link with reproduction steps isn't accessible. Can you post the reproduction steps? Thanks!

**Comment by brophdawg11** [comment]
The reproduction is here but I'm not sure it quite reproduces on stackblitz because they serve JS files from a service worker so we don't hit the underlying race condition: https://stackblitz.com/edit/remix-run-remix-qdy2d2

**Comment by cmerrick** [comment]
Confirmed I was able to reproduce using that code once I activated Network Throttling in chrome (set to 'Slow 3G'). Error looks like this:

<img width="772" alt="image" src="https://github.com/remix-run/react-router/assets/207186/c1ee1361-9de1-42b7-8985-cdac5bd10d48">

**Comment by n8agrin** [comment]
@brophdawg11 I hacked together a "fix" in my fork of Remix: https://github.com/n8agrin/remix/pull/1

The pattern I'm noticing is that in many contexts (shouldRevalidate being just one), Remix attempts to access routeModules assuming that all possible required modules are always loaded. We're seeing that's not always the case in this issue and in the issue I'm trying to fix in this other PR: https://github.com/remix-run/remix/pull/6409

My unsolicited 2c is that access to routeModules should be behind an interface that can: retrieve from local cache, load from the network or throw.

I'm trying to begin to see what that would feel like in this pr, and I don't love how the async nature of the problem demands more and more of the code be refactored to manage one small dependency.

Note that I tried the reload approach (as in https://github.com/remix-run/remix/pull/6409), but that doesn't work. It creates an infinite loop within the shouldRevalide call (no surprise I suppose).

Do you have any ideas? Maybe we need to bite the bullet and do the async refactor.

**Comment by n8agrin** [comment]
@brophdawg11 just realized some of my edits span react-router and they aren’t captured here. I’ll upload the diff later. In particular I changed the createShouldRevalidate generator to async load a route module if it’s not found, which solves this issue

**Comment by brophdawg11** [comment]
@n8agrin I don't think we want to make `shouldRevalidate` async - it's intended to be a synchronous decision on whether or not to proceed with the async revalidation.  The bug here wasn't that we expected the route module to be loaded - it was instead that this wasn't a "revalidation" at all since no initial load have ever completed, so it's just an initial load and doesn't need to be checking `shouldRevalidate` at all.

**Comment by n8agrin** [comment]
@brophdawg11 let me try to explain the bug we're seeing. It's not readily clear that the fix for this issue will fix our particular problem:

* occasionally, loaders will return undefined
* we possibly connected the problem to this issue. We're not 100% certain, but it has similar symptoms
* when i traced the code via my repro case the code seems to be trying to load a route module that is not in the cache
* this seems to tie into the shouldRevalidate logic
* I believe it's not the first time the route logic for the loader is exercised. Our e2e test that hits this issue tends to trigger it deep in the test, well after other actions should have triggered the same loader

If I understand your patch correctly, it doesn't trigger shouldRevalidate on initial route load, because the routeModule may not yet be loaded. My mental model about how this works is confused, because it seems like the synchronous logic assumes all of the route modules must be loaded, when they clearly may not be.

Two questions:
1. Based on the description of the behavior of our bug, do you think this patch will fix the issue? (I will also attempt to test on a build with this patch)
2. Can you help me refine my mental model about how Router / Remix handle route modules, and in particular the interaction between synchronous code paths managing assets available asynchronously?

**Comment by n8agrin** [comment]
@brophdawg11 awesome thanks for that

**Comment by n8agrin** [comment]
@brophdawg11 this didn't seem to fix our issue. I'll try to craft a repro, post it and/or provide a patch
